astropy/time/formats.py

```diff
@@ -1539,24 +1539,28 @@ def _check_val_type(self, val1, val2):
 
     def set_jds(self, val1, val2):
         self._check_scale(self._scale)  # Validate scale.
-        iterator = np.nditer([val1, None],
+        iterator = np.nditer([val1, None, None],
                              flags=['refs_ok', 'zerosize_ok'],
-                             op_dtypes=[None] + [np.double])
+                             op_dtypes=[None, np.double, np.double])
 
-        for val, sec in iterator:
-            sec[...] = val.item().total_seconds()
+        day = datetime.timedelta(days=1)
+        for val, jd1, jd2 in iterator:
+            jd1[...], other = divmod(val.item(), day)
+            jd2[...] = other/day
 
-        self.jd1, self.jd2 = day_frac(iterator.operands[-1], 0.0,
-                                      divisor=erfa.DAYSEC)
+        self.jd1, self.jd2 = day_frac(iterator.operands[-2],
+                                      iterator.operands[-1])
 
     @property
     def value(self):
-        iterator = np.nditer([self.jd1 + self.jd2, None],
+        iterator = np.nditer([self.jd1, self.jd2, None],
                              flags=['refs_ok', 'zerosize_ok'],
-                             op_dtypes=[None] + [object])
+                             op_dtypes=[None, None, object])
 
-        for jd, out in iterator:
-            out[...] = datetime.timedelta(days=jd.item())
+        for jd1, jd2, out in iterator:
+            jd1_, jd2_ = day_frac(jd1.item(), jd2.item())
```

Comment:
The `.item()` shouldn't be necessary here. Also, would *think* that the `day_frac` is not needed either, since `timedelta` would do it internally (but haven't tested).

Reply:
The existing code used .item() but I don't know why, so I left it.

Didn't we have this discussion, that `astropy` does not guarantee anything about the values of `jd1` and `jd2`? Specifically, custom formats pass through no checking, and only some code will break when presented with unusual values here. Scale conversion could also produce non-compliant `jd1` and `jd2`, though I think the fix for that was merged. My argument was that having a near-invariant was worse than not providing any invariant, because code would depend on it, as your suggestion does, and then some obscure bits of code would break, in this case in subtle ways.

Reply:
Yes, we had. My suggestion was simply to check whether `timedelta` does the rounding for one and thus takes care of the case where `jd1` and `jd2` are no longer quite as expected. And it does in fact do this, so I guess here one is mostly guarding against the case where jd1 and jd2 would be swapped or so.
In any case, it doesn't really matter - an `object` array of `timedelta` is going to be so slow to deal with that an extra call to `day_frac` will not add much.

Reply:
p.s. Let's remove the `.item()` - I'm fairly sure it is a copy&paste mistake from the other direction, where, since one deals with objects, it makes sense to get the actual `timedelta` out.

Reply:
I sort of want to create a custom format that constructs non-normalized jd1/jd2 and see how much stuff subtly breaks. The ERFA test case in `test_sidereal` is one example, this suggested implementation is another, I wonder how much else is baked in to our code base? Not everything, most stuff would be fine.